mod/quiz/classes/external/add_random_question.php

```diff
@@ -0,0 +1,117 @@
+                'parentcategory' => new external_value(PARAM_TEXT, 'Parent category of new category', VALUE_DEFAULT, 0),
+            ]
+        );
+    }
+
+    /**
+     *  Add random questions.
+     *
+     * @param int $cmid The cmid of the quiz
+     * @param int $categoryid Question category
+     * @param int $addonpage The page where random questions will be added to
+     * @param int $randomcount Number of random questions
+     * @param string $filtercondition Filter condition
+     * @param string $newcategory add new category
+     * @param string $parentcategory parent category of new category
+     * @return array result
+     */
+    public static function execute(int $cmid, int $categoryid, int $addonpage, int $randomcount,
+                                   string $filtercondition, string $newcategory = '', string $parentcategory = ''): array {
+        [
+            'cmid' => $cmid,
+            'categoryid' => $categoryid,
+            'addonpage' => $addonpage,
+            'randomcount' => $randomcount,
+            'filtercondition' => $filtercondition,
+            'newcategory' => $newcategory,
+            'parentcategory' => $parentcategory,
+        ] = self::validate_parameters(self::execute_parameters(), [
+            'cmid' => $cmid,
+            'categoryid' => $categoryid,
+            'addonpage' => $addonpage,
+            'randomcount' => $randomcount,
+            'filtercondition' => $filtercondition,
+            'newcategory' => $newcategory,
+            'parentcategory' => $parentcategory,
+        ]);
+
+        // Validate context.
+        $thiscontext = \context_module::instance($cmid);
+        self::validate_context($thiscontext);
+        require_capability('mod/quiz:manage', $thiscontext);
+
+        // Create new category.
+        if (!empty($newcategory)) {
+            $contexts = new \core_question\local\bank\question_edit_contexts($thiscontext);
+            $defaultcategoryobj = question_make_default_categories($contexts->all());
+            $defaultcategory = $defaultcategoryobj->id . ',' . $defaultcategoryobj->contextid;
+            $qcobject = new \qbank_managecategories\question_category_object(
+                null,
+                null,
+                $contexts->having_one_edit_tab_cap('categories'),
+                $defaultcategoryobj->id,
+                $defaultcategory,
+                null,
+                $contexts->having_cap('moodle/question:add'));
+            $categoryid = $qcobject->add_category($parentcategory, $newcategory, '', true);
+        }
+
+        // Add random question to the quiz.
+        list($quiz, ) = get_module_from_cmid($cmid);
+        quiz_add_random_questions($quiz, $addonpage, $categoryid, $randomcount, $filtercondition);
+        quiz_delete_previews($quiz);
+        quiz_update_sumgrades($quiz);
+
+        return ['message' => get_string('addarandomquestion_success', 'mod_quiz')];
```

Comment:
is this used in ui? cant see that as a user, please lemme know if I am missing something

Reply:
The ws function is used in this js: https://github.com/catalyst/moodle-MDL-72752/pull/63/files#diff-ad78a10007735f9361ff037bfec8321d5502d65036d064c6a3628692b3cc4729R334

Reply:
correct, but why the return message? the return message seems not in use

Reply:
I have updated with messages are used in https://github.com/catalyst/moodle-MDL-72752/pull/63/commits/fc6dea0aef23893438f9ef3da241f745d296077b#diff-9cdb99b99ca28cfcf036270f6b963eb974353a52945cffb6ae4752454d6702b9R173